File: specs/typespec/employer/suborgs.ts
import {
	type ValidationError,
	newValidationError,
	ERR_REQUIRED,
} from "../common/common";

export type SubOrgStatus = "active" | "disabled";

export const VALID_SUBORG_REGIONS = ["ind1", "usa1", "deu1", "sgp1"] as const;
export type SubOrgRegion = (typeof VALID_SUBORG_REGIONS)[number];

export interface SubOrg {
	id: string;
	name: string;
	pinned_region: string;
	status: SubOrgStatus;
	created_at: string;
}

export interface SubOrgMember {
	email_address: string;
	name: string;
	assigned_at: string;
}

export interface CreateSubOrgRequest {
	name: string;
	pinned_region: string;
}

export interface ListSubOrgsRequest {
	filter_status?: SubOrgStatus;
	cursor?: string;
	limit?: number;
}

export interface ListSubOrgsResponse {
	suborgs: SubOrg[];
	next_cursor: string;
}

export interface RenameSubOrgRequest {
	suborg_id: string;
	name: string;
}

export interface DisableSubOrgRequest {
	suborg_id: string;
}

export interface EnableSubOrgRequest {
	suborg_id: string;
}

export interface AddSubOrgMemberRequest {
	suborg_id: string;
	email_address: string;
}

export interface RemoveSubOrgMemberRequest {
	suborg_id: string;
	email_address: string;
}

export interface ListSubOrgMembersRequest {
	suborg_id: string;
	cursor?: string;
	limit?: number;
}

export interface ListSubOrgMembersResponse {
	members: SubOrgMember[];
	next_cursor: string;
}

const ERR_NAME_TOO_LONG = "name must be at most 64 characters";
const ERR_REGION_INVALID =
	"pinned_region must be one of: ind1, usa1, deu1, sgp1";
const ERR_FILTER_STATUS_INVALID =
	"filter_status must be 'active' or 'disabled'";

export function validateCreateSubOrgRequest(
	request: CreateSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.name) {
		errs.push(newValidationError("name", ERR_REQUIRED));
	} else if (request.name.length > 64) {
		errs.push(newValidationError("name", ERR_NAME_TOO_LONG));
	}

	if (!request.pinned_region) {
		errs.push(newValidationError("pinned_region", ERR_REQUIRED));
	} else if (
		!VALID_SUBORG_REGIONS.includes(request.pinned_region as SubOrgRegion)
	) {
		errs.push(newValidationError("pinned_region", ERR_REGION_INVALID));
	}

	return errs;
}

export function validateListSubOrgsRequest(
	request: ListSubOrgsRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (
		request.filter_status &&
		request.filter_status !== "active" &&
		request.filter_status !== "disabled"
	) {
		errs.push(newValidationError("filter_status", ERR_FILTER_STATUS_INVALID));
	}

	return errs;
}

export function validateRenameSubOrgRequest(
	request: RenameSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.suborg_id) {
		errs.push(newValidationError("suborg_id", ERR_REQUIRED));
	}

	if (!request.name) {
		errs.push(newValidationError("name", ERR_REQUIRED));
	} else if (request.name.length > 64) {
		errs.push(newValidationError("name", ERR_NAME_TOO_LONG));
	}

	return errs;
}

export function validateDisableSubOrgRequest(
	request: DisableSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!request.suborg_id) {
		errs.push(newValidationError("suborg_id", ERR_REQUIRED));
	}
	return errs;
}

export function validateEnableSubOrgRequest(
	request: EnableSubOrgRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!request.suborg_id) {
		errs.push(newValidationError("suborg_id", ERR_REQUIRED));
	}
	return errs;
}

export function validateAddSubOrgMemberRequest(
	request: AddSubOrgMemberRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!request.suborg_id) {
		errs.push(newValidationError("suborg_id", ERR_REQUIRED));
	}
	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_REQUIRED));
	}
	return errs;
}

export function validateRemoveSubOrgMemberRequest(
	request: RemoveSubOrgMemberRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!request.suborg_id) {
		errs.push(newValidationError("suborg_id", ERR_REQUIRED));
	}
	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_REQUIRED));
	}
	return errs;
}

export function validateListSubOrgMembersRequest(
	request: ListSubOrgMembersRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!request.suborg_id) {
		errs.push(newValidationError("suborg_id", ERR_REQUIRED));
	}
	return errs;
}

File: specs/typespec/employer/suborgs.go
package employer

import (
	"fmt"

	"vetchium-api-server.typespec/common"
)

const (
	subOrgNameMaxLength = 64

	errSubOrgNameRequired   = "name is required"
	errSubOrgNameTooLong    = "name must be at most 64 characters"
	errSubOrgIDRequired     = "suborg_id is required"
	errSubOrgRegionRequired = "pinned_region is required"
	errSubOrgEmailRequired  = "email_address is required"
)

// SubOrg is the response type for SubOrg reads.
type SubOrg struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	PinnedRegion string `json:"pinned_region"`
	Status       string `json:"status"`
	CreatedAt    string `json:"created_at"`
}

// SubOrgMember is a member of a SubOrg.
type SubOrgMember struct {
	EmailAddress string `json:"email_address"`
	Name         string `json:"name"`
	AssignedAt   string `json:"assigned_at"`
}

// CreateSubOrgRequest is the request body for POST /employer/create-suborg.
type CreateSubOrgRequest struct {
	Name         string `json:"name"`
	PinnedRegion string `json:"pinned_region"`
}

func (r CreateSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	} else if len(r.Name) > subOrgNameMaxLength {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameTooLong)))
	}

	if r.PinnedRegion == "" {
		errs = append(errs, common.NewValidationError("pinned_region", fmt.Errorf(errSubOrgRegionRequired)))
	}

	return errs
}

// ListSubOrgsRequest is the request body for POST /employer/list-suborgs.
type ListSubOrgsRequest struct {
	FilterStatus *string `json:"filter_status,omitempty"`
	Cursor       *string `json:"cursor,omitempty"`
	Limit        *int32  `json:"limit,omitempty"`
}

func (r ListSubOrgsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.FilterStatus != nil && *r.FilterStatus != "active" && *r.FilterStatus != "disabled" {
		errs = append(errs, common.NewValidationError("filter_status", fmt.Errorf("filter_status must be 'active' or 'disabled'")))
	}

	return errs
}

// ListSubOrgsResponse is the response for POST /employer/list-suborgs.
type ListSubOrgsResponse struct {
	SubOrgs    []SubOrg `json:"suborgs"`
	NextCursor string   `json:"next_cursor"`
}

// RenameSubOrgRequest is the request body for POST /employer/rename-suborg.
type RenameSubOrgRequest struct {
	SubOrgID string `json:"suborg_id"`
	Name     string `json:"name"`
}

func (r RenameSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SubOrgID == "" {
		errs = append(errs, common.NewValidationError("suborg_id", fmt.Errorf(errSubOrgIDRequired)))
	}

	if r.Name == "" {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameRequired)))
	} else if len(r.Name) > subOrgNameMaxLength {
		errs = append(errs, common.NewValidationError("name", fmt.Errorf(errSubOrgNameTooLong)))
	}

	return errs
}

// DisableSubOrgRequest is the request body for POST /employer/disable-suborg.
type DisableSubOrgRequest struct {
	SubOrgID string `json:"suborg_id"`
}

func (r DisableSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SubOrgID == "" {
		errs = append(errs, common.NewValidationError("suborg_id", fmt.Errorf(errSubOrgIDRequired)))
	}

	return errs
}

// EnableSubOrgRequest is the request body for POST /employer/enable-suborg.
type EnableSubOrgRequest struct {
	SubOrgID string `json:"suborg_id"`
}

func (r EnableSubOrgRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SubOrgID == "" {
		errs = append(errs, common.NewValidationError("suborg_id", fmt.Errorf(errSubOrgIDRequired)))
	}

	return errs
}

// AddSubOrgMemberRequest is the request body for POST /employer/add-suborg-member.
type AddSubOrgMemberRequest struct {
	SubOrgID     string `json:"suborg_id"`
	EmailAddress string `json:"email_address"`
}

func (r AddSubOrgMemberRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SubOrgID == "" {
		errs = append(errs, common.NewValidationError("suborg_id", fmt.Errorf(errSubOrgIDRequired)))
	}

	if r.EmailAddress == "" {
		errs = append(errs, common.NewValidationError("email_address", fmt.Errorf(errSubOrgEmailRequired)))
	}

	return errs
}

// RemoveSubOrgMemberRequest is the request body for POST /employer/remove-suborg-member.
type RemoveSubOrgMemberRequest struct {
	SubOrgID     string `json:"suborg_id"`
	EmailAddress string `json:"email_address"`
}

func (r RemoveSubOrgMemberRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SubOrgID == "" {
		errs = append(errs, common.NewValidationError("suborg_id", fmt.Errorf(errSubOrgIDRequired)))
	}

	if r.EmailAddress == "" {
		errs = append(errs, common.NewValidationError("email_address", fmt.Errorf(errSubOrgEmailRequired)))
	}

	return errs
}

// ListSubOrgMembersRequest is the request body for POST /employer/list-suborg-members.
type ListSubOrgMembersRequest struct {
	SubOrgID string  `json:"suborg_id"`
	Cursor   *string `json:"cursor,omitempty"`
	Limit    *int32  `json:"limit,omitempty"`
}

func (r ListSubOrgMembersRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SubOrgID == "" {
		errs = append(errs, common.NewValidationError("suborg_id", fmt.Errorf(errSubOrgIDRequired)))
	}

	return errs
}

// ListSubOrgMembersResponse is the response for POST /employer/list-suborg-members.
type ListSubOrgMembersResponse struct {
	Members    []SubOrgMember `json:"members"`
	NextCursor string         `json:"next_cursor"`
}
